language-support/scala/bindings-akka/src/main/scala/com/digitalasset/ledger/client/binding/retrying/CommandRetryFlow.scala

```diff
@@ -72,12 +72,12 @@ object CommandRetryFlow {
                 } else if ((firstSubmissionTime plus maxRetryTime) isBefore timeProvider.getCurrentTime) {
                   RetryLogger.logStopRetrying(request, status, nrOfRetries, firstSubmissionTime)
                   PROPAGATE_PORT
-                } else if (NON_RETRYABLE_ERROR_CODES.contains(status.code)) {
-                  RetryLogger.logFatal(request, status, nrOfRetries)
-                  PROPAGATE_PORT
-                } else {
+                } else if (RETRYABLE_ERROR_CODES.contains(status.code)) {
```

Comment:
Does this logic operate only on submission or on both submission and completion? Would this result in bugs if a completion would start to return an error code in `RETRYABLE_ERROR_CODES`?

Reply:
This operates on whatever produces a `Completion` value. Unfortunately the client side command tracking mechanism produces such values it self, even though it may not have received it from the ledger. This is used to communicate timeouts (ABORTED) and backpressure (RESOURCE_EXHAUSTED). The reason why I didn't count the ABORTED status code as retryable is that ABORTED is also used for actual rejections in completions.

Reply:
It's probably better to remove it altogether, because

- it's very brittle
- only works for *very* few cases
- the application needs to handle failure anyway, because even the retry mechanism can run into an upper bound of retries and the command can still fail.

Reply:
I agree with @gerolf-da, application should handle failures/retries.
The change looks good, does what it supposed to. I was actually going to ask about `ABORTED` and `UNAVAILABLE`... `ABORTED` case is explained above, how about `UNAVAILABLE`?
Here is the link explaining GRPC error codes: https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
```
The service is currently unavailable. This is most likely a transient condition, which can be corrected by retrying with a backoff. Note that it is not always safe to retry non-idempotent operations.
```
I think it makes sense to add `UNAVAILABLE` to `RETRYABLE_ERROR_CODES`.

Reply:
do we actually claim that command submission is `idempotent`?

Reply:
I guess we don't claim it at all, read the PR summary:

>  Retrying completion failures with the same
commandId doesn't make sense, as the deduplication mechanism will kick
in.

What happens when deduplication mechanism kicks off?

Reply:
In case of a transaction, it gets rejected. But interestingly, the deduplication mechanism currently doesn't work on rejections. In that case storing the rejection fails with a sql exception.